[WIP] Delete project from disk

melanieseltzer opened the pull request:
**Related Issue:**
#90 and #140

**Summary:**
This is the functionality for deleting project from disk. Right now it lives in the Project app menu but will eventually also live in the modal from #120. And eventually it will also have 'Delete from Guppy' option as well. It will also need to be converted to a saga once #154 lands.

Right now things seem to be 99% working, but there's an annoying bug I can't seem to figure out where the circular animations are not firing consistently. If you delete a project that's anything but the first in the list it will run, but not the first in the list 😕

In this PR I've also tweaked the Project menu conditional slightly. It was originally only showing up if `selectedProjectId && devServerTask` and this was fine when it just had the clear log task, but now it needs to always be there since user will always be able to delete a project. I noticed this issue because some of my projects don't have a start task (so, no `devServerTask` either) so there's no `Development Server` pane, and the Project menu wasn't showing up at all.

**Screenshots/GIFs:**

Normal view:

<img width="772" alt="screen shot 2018-08-26 at 9 01 28 am" src="https://user-images.githubusercontent.com/17421347/44630214-d7cd6a80-a90e-11e8-8d50-20cf3d39da9c.png">

No start task:

<img width="769" alt="screen shot 2018-08-26 at 9 01 59 am" src="https://user-images.githubusercontent.com/17421347/44630210-cc7a3f00-a90e-11e8-931f-5bf81470fe81.png">

Circular animation bug:

![kapture 2018-08-26 at 9 07 39](https://user-images.githubusercontent.com/17421347/44630294-d2245480-a90f-11e8-8c29-147f8d3a8977.gif)

j-f1 commented on src/middlewares/delete-project.middleware.js:
```diff
@@ -0,0 +1,86 @@
+      // If remainingProjects array has at least one project
+      // then there's a project to select next
+      if (remainingProjects.length >= 1) {
+        const nextProject = remainingProjects[0];
```
Perhaps try to find the project directly after the deleted project (or, if it’s the last one, the one before it)?

AWolf81 replied:
Yes, I also think selecting the next project would be better. 
Something like `const index = projects.indexOf(deletedProject); const nextIndex = index < remainigProjects.length ? index : 0` could work.
The code is untested but it should move the index to the next project.

joshwcomeau replied:
Right, so I think I was the one who suggested going for the next one.

I don't have strong opinions. In fact, I think in the future the two options will be the same (I think I want to sort the projects by recently-accessed, so selecting a project would move it to the top spot).

But yeah, I don't have a preference for what we do in the meantime :)

AWolf81 commented on src/middlewares/delete-project.middleware.js:
```diff
@@ -0,0 +1,86 @@
+
+      if (successfullyDeleted) {
+        store.dispatch(deleteProjectFromDiskFinish(project));
+        store.dispatch(refreshProjects());
```
I think dispatching `refreshProjects` is not needed here if you're adding the case `FINISH_DELETING_PROJECT_FROM_DISK` to `projects.reducer` and remove the deleted project from state there.

I think I had it similar in project configuration change and refresh created some troubles - it's possible that this will improve the selection issue.

AWolf81 commented on src/components/ApplicationMenu/ApplicationMenu.js:
```diff
@@ -154,19 +159,29 @@ class ApplicationMenu extends Component<Props> {
 
+      // If there's no devServerTask then hide that menu item
+      // this will happen if the project doesn't have any start tasks
+      const clearConsoleMenuItems = devServerTask
```
Nit. I think you could write `devServerTask && { ... menuItem code }` here - this will return `undefined` if there is no devServer otherwise it will add the item.

joshwcomeau replied:
I _believe_ it complains if you try to pass `undefined` to a submenu; it doesn't filter out falsy items.

Instead I think I'd do something like this:

```js
let submenu = [];

if (devServerTask) {
  submenu.push(/* clear console item */);
}

submenu.push(/* delete menu item */);
```

And then just use `submenu` in the splice.

joshwcomeau commented on src/actions/index.js:
```diff
@@ -56,6 +60,21 @@ export const addProject = (project: Project) => ({
+export const deleteProjectFromDisk = (project: string) => ({
+  type: DELETE_PROJECT_FROM_DISK,
+  project,
+});
```
I don't believe we need this action; see below for more info :)

joshwcomeau commented on src/actions/index.js:
```diff
@@ -56,6 +60,21 @@ export const addProject = (project: Project) => ({
+});
+
+export const deleteProjectFromDiskFinish = (deletedProject: string) => ({
+  type: FINISH_DELETING_PROJECT_FROM_DISK,
```
Looks like there's an inconsistency between the action-creator name and the action constant.

Since I think we should kill `DELETE_PROJECT_FROM_DISK`, I quite like `FINISH_DELETING_PROJECT_FROM_DISK`; I think we should update the action-creator name to be the same :)

joshwcomeau commented on src/actions/index.js:
```diff
@@ -56,6 +60,21 @@ export const addProject = (project: Project) => ({
+
+export const deleteProjectFromDiskFinish = (deletedProject: string) => ({
+  type: FINISH_DELETING_PROJECT_FROM_DISK,
+  deletedProject,
```
Couple nits about this variable:

- I think we should include the least amount of info with an action, and in this case, I think we could get away with just the `projectId`, rather than the whole project.
- I like when names of data entities are consistent, so I'd name this `project` instead of `deletedProject` (or `projectId` if we can just use the ID)

joshwcomeau commented on src/middlewares/delete-project.middleware.js:
```diff
@@ -0,0 +1,86 @@
+
+          if (shouldDeleteFromDisk) {
+            const { project } = action;
+            store.dispatch(deleteProjectFromDisk(project));
```
I think we can kill the bottom 2 cases :) this'll set up up well for the saga.

Essentially it's just a matter of squashing the 3 together, and omitting the `DELETE_PROJECT_FROM_DISK` dispatch altogether.


```js
case SHOW_DELETE_PROJECT_PROMPT:
  dialog.showMessageBox(..., response => {
    if (!shouldDeleteFromDisk) {
      return;
    }

    const successfullyDeleted = shell.moveItemToTrash(path);

    if (successfullyDeleted) {
      store.dispatch(deleteProjectFromDiskFinish(project));
    }
    // code to dispatch either selectProject or createNewProjectWizard
  })
```

I agree with @AWolf81's comment below, I don't think we need `refreshProjects`. Instead I'd update the projects reducer to remove the project.

I did this in a local copy of your branch, and the code looks like this (in `projects.reducer.js`):

```js
    case FINISH_DELETING_PROJECT_FROM_DISK: {
      const { deletedProject } = action;

      return produce(state, draftState => {
        delete draftState[deletedProject.id];
      });
    }
```

(note that it would just use `projectId` instead of `deletedProject.id` if you follow my earlier comment)

melanieseltzer commented:
Thank you for the great feedback everyone 👌 I have implemented all the suggested changes and it looks like we are in much better shape than before.

- Implemented @j-f1's and @AWolf81's next project suggestion. Next project that gets selected is now either the one after the deleted project, or if the deleted project was at the end of the array then the one before it is selected.
- Implemented @joshwcomeau's suggestions.

Now with these changes, the circular animation runs on every deletion, *except* if the deletion is the first element in the array (position 0). I'm wondering if there is something significant with that zero number causing the animation to fail.

The one last snag I've noticed, is that upon mount `devServerTask` is returning null 🤷‍♀️ Once you select a project by clicking on the icon, the variable is properly set.  But the application menu render is firing too quickly so it's catching null, and the clear console log option isn't getting pushed into the array. 

For now I did a hacky fix with some setTimeouts but obviously this isn't ideal. Any thoughts?

AWolf81 commented on src/middlewares/delete-project.middleware.js:
```diff
@@ -0,0 +1,83 @@
+          const nextIndex =
+            index < remainingProjects.length
+              ? index
+              : remainingProjects.length - 1;
```
Good idea to select the last one. :+1:
I like it more then selecting the first one but I think both would work.

joshwcomeau replied:
Agreed!

Nit: I think `nextIndex` is only used if there are remaining projects, so it could be moved under the `if (remainingProjects.length >= `)` conditional

joshwcomeau commented on src/components/ApplicationMenu/ApplicationMenu.js:
```diff
@@ -21,18 +22,22 @@ import type { Task } from '../../types';
-    this.buildMenu();
+    setTimeout(() => {
+      this.buildMenu();
+    }, 1000);
```
I noticed this issue myself, during testing. I'll take a look after this lands :) @AWolf81 has some good ideas

joshwcomeau commented on src/components/ApplicationMenu/ApplicationMenu.js:
```diff
@@ -154,21 +163,31 @@ class ApplicationMenu extends Component<Props> {
       // have a different index depending on the platform.
       const editMenuIndex = template.findIndex(menu => menu.id === 'edit');
 
+      // Only include clear console menu item if devServerTask exists
```
Nit: this comment should be for the `if` statement 2 lines down

joshwcomeau commented on src/components/ApplicationMenu/ApplicationMenu.js:
```diff
@@ -154,21 +163,31 @@ class ApplicationMenu extends Component<Props> {
+
       template.splice(editMenuIndex, 0, {
         id: 'project',
         label: isMac ? 'Project' : '&Project',
```
While I was testing your changes locally, I found I kept expecting "Create new project" to be within the `Project` menu.

Unrelated to this PR, and non-blocking, but what do you think about us renaming this to `Current Project`? That way it's clear that this menu is meant to deal with the currently-selected project?

melanieseltzer replied:
👍 Yeah I agree, I think it would be more descriptive.

joshwcomeau commented:
> [Edit] Wait a second, could the issue with devServerTask being null be related to your issue #163, @joshwcomeau? Since you mentioned in that PR the dev server pane isn't immediately available.

Yep! It's the same issue, and my fix in #164 will fix this as well (just tested locally) :D

melanieseltzer commented:
🎉 Yay it landed! Thanks everyone!

@joshwcomeau I'll submit a quick PR for those lingering things.
